dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/WorkflowEventBusFireWorker.java

```diff
@@ -0,0 +1,140 @@
+        final Integer workflowInstanceId = workflowInstance.getId();
+        registeredWorkflowExecuteRunnableMap.remove(workflowInstanceId, workflowExecutionRunnable);
+    }
+
+    public void fireAllRegisteredEvent() {
+        final List<IWorkflowExecutionRunnable> workflowExecutionRunnables = getWaitingFireWorkflowExecutionRunnables();
+        if (CollectionUtils.isEmpty(workflowExecutionRunnables)) {
+            return;
+        }
+        for (IWorkflowExecutionRunnable workflowExecutionRunnable : workflowExecutionRunnables) {
+            final Integer workflowInstanceId = workflowExecutionRunnable.getId();
+            final String workflowInstanceName = workflowExecutionRunnable.getName();
+            try {
+                LogUtils.setWorkflowInstanceIdMDC(workflowInstanceId);
+                doFireSingleWorkflowEventBus(workflowExecutionRunnable);
+            } catch (Exception ex) {
+                log.error("Fire event failed for WorkflowExecuteRunnable: {}", workflowInstanceName, ex);
+            } finally {
+                LogUtils.removeWorkflowInstanceIdMDC();
+            }
+        }
+    }
+
+    private List<IWorkflowExecutionRunnable> getWaitingFireWorkflowExecutionRunnables() {
+        if (MapUtils.isEmpty(registeredWorkflowExecuteRunnableMap)) {
+            return Collections.emptyList();
+        }
+        return registeredWorkflowExecuteRunnableMap.values()
+                .stream()
+                .filter(workflowExecuteRunnable -> !workflowExecuteRunnable.getWorkflowEventBus().isEmpty())
+                .collect(Collectors.toList());
+    }
+
+    private void doFireSingleWorkflowEventBus(final IWorkflowExecutionRunnable workflowExecutionRunnable) {
+        final WorkflowEventBus workflowEventBus = workflowExecutionRunnable.getWorkflowEventBus();
+        while (!workflowEventBus.isEmpty()) {
+            Optional<AbstractLifecycleEvent> eventOptional = workflowEventBus.poll();
+            if (!eventOptional.isPresent()) {
+                return;
+            }
+            final AbstractLifecycleEvent lifecycleEvent = eventOptional.get();
+            try {
+                // Since we will print the event count at FinalizeEventHandler
+                // So we increase the event count before the event fired then we can get the correct event count
+                // And if the event handle failed we will decrease the success event count
+                workflowEventBus.getWorkflowEventBusSummary().increaseFireSuccessEventCount();
+                doFireSingleEvent(workflowExecutionRunnable, lifecycleEvent);
+            } catch (Exception ex) {
+                // If the database connection is failed, do not remove the event from the event bus
+                // so that the event can be fired again when the database connection is recovered
+                if (ExceptionUtils.isDatabaseConnectedFailedException(ex)) {
+                    workflowEventBus.publish(lifecycleEvent);
```

Comment:
Should return here?  Better to sleep when db connect fail.

Reply:
I added sleep here.